File: src/main/java/com/ticketing/domain/reservation/Reservation.java
package com.ticketing.domain.reservation;

import com.ticketing.domain.concert.Concert;
import com.ticketing.domain.member.Member;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "reservations", indexes = {
        @Index(name = "idx_request_id", columnList = "request_id", unique = true)
})
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Reservation {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "request_id", unique = true, nullable = false, length = 36)
    private String requestId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    private Member member;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "concert_id", nullable = false)
    private Concert concert;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReservationStatus status;

    public Reservation(String requestId, Member member, Concert concert, ReservationStatus status) {
        this.requestId = requestId;
        this.member = member;
        this.concert = concert;
        this.status = status;
    }

    // 상태 업데이트 메서드
    public void complete() {
        this.status = ReservationStatus.COMPLETED;
    }

    public void fail() {
        this.status = ReservationStatus.FAILED;
    }

    public void cancel() {
        this.status = ReservationStatus.CANCELLED;
    }
}

File: src/main/kotlin/com/ticketing/domain/reservation/Reservation.kt
package com.ticketing.domain.reservation

import com.ticketing.domain.concert.Concert
import com.ticketing.domain.member.Member
import com.ticketing.domain.ticket.Ticket
import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
@Table(name = "reservations")
data class Reservation(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    val member: Member,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "concert_id", nullable = false)
    val concert: Concert,

    @OneToMany(mappedBy = "reservation", cascade = [CascadeType.ALL], orphanRemoval = true)
    val tickets: MutableList<Ticket> = mutableListOf(),

    @Column(nullable = false)
    val ticketCount: Int,

    @Column(nullable = false)
    val totalAmount: BigDecimal,

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    val status: ReservationStatus = ReservationStatus.PENDING,

    @Column(nullable = false, unique = true, length = 50)
    val reservationNumber: String,

    @Column(nullable = false, updatable = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(nullable = false)
    var updatedAt: LocalDateTime = LocalDateTime.now(),

    @Column
    var expiredAt: LocalDateTime? = null
) {
    fun confirm() {
        // 상태는 외부에서 변경
    }

    fun cancel() {
        // 취소 로직
    }

    fun isExpired(): Boolean {
        return expiredAt != null && expiredAt!!.isBefore(LocalDateTime.now())
    }
}

enum class ReservationStatus {
    PENDING, CONFIRMED, CANCELLED, EXPIRED
}

